### ISSUE TITLE
CombinedLoader in DDP draw the same sample multiple times in one epoch

### ISSUE THREAD

**Issue by wuchiz** [Lightning-AI/pytorch-lightning#19188]
### Bug description

When I trained model with multiple dataloaders using CombinedLoader (gpu=2), the loss increased, while loss normally decreased when I use only one gpu. Then I checked the index of sample finding that dataloader in different process gave the same sample.

### What version are you seeing the problem on?

v2.1

### How to reproduce the bug

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from lightning.pytorch import LightningModule, Trainer, LightningDataModule
from lightning.pytorch.utilities import CombinedLoader


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        print(f"index:{index}")
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        # print(f"Training rank {self.trainer.local_rank} index {batch_idx}")
        loss = self(batch["a"]).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        # print(f"Validation rank {self.trainer.local_rank} index {batch_idx}")
        loss = self(batch["a"]).sum()
        self.log("valid_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)

len_a = 10
len_b = 6
mode = "min_size"
class MyData(LightningDataModule):

    def train_dataloader(self):
        return CombinedLoader( {"a": DataLoader(RandomDataset(32, len_a), batch_size=1,num_workers=0), 
        "b": DataLoader(RandomDataset(32, len_b), batch_size=1,num_workers=0)}, mode=mode)

    def val_dataloader(self):
        return CombinedLoader({"a": DataLoader(RandomDataset(32, len_a), batch_size=1), 
        "b": DataLoader(RandomDataset(32, len_b), batch_size=1)}, mode=mode)

def run():
    
    model = BoringModel()
    print("starting")
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        max_epochs=2,
        accelerator="cuda",
        devices=2,
        strategy='ddp',
        num_sanity_val_steps=0,
    )
    trainer.fit(model, datamodule=MyData())


if __name__ == "__main__":
    run()
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>
